File: Examples/GroupDocs.Annotation.VisualBasic/GroupDocs.AnnotationDataStorage/Repositories/AnnotationReplyRepository.vb
Imports System.Collections.Generic
Imports System.Linq
Imports GroupDocs.Annotation.Handler.Input.DataObjects
Imports GroupDocs.Annotation.Handler.Input
Namespace GroupDocs.Data.Json.Repositories
    Public Class AnnotationReplyRepository
        Inherits JsonRepository(Of AnnotationReply)
        Implements IAnnotationReplyDataHandler

        Private Const _repoName As String = "GroupDocs.annotation.replies.json"

        Public Sub New(pathFinder As IRepositoryPathFinder)
            MyBase.New(pathFinder.Find(_repoName))
        End Sub

        Public Sub New(filePath As String)
            MyBase.New(filePath)
        End Sub

        Public Function GetReply(guid As String) As AnnotationReply Implements IAnnotationReplyDataHandler.GetReply
            SyncLock _syncRoot
                Try
                    Return Data.FirstOrDefault(Function(x) x.Guid = guid)
                Catch e As Exception
                    Throw New DataJsonException("Failed to get annotation reply by GUID.", e)
                End Try
            End SyncLock
        End Function

        Public Function GetReplies(annotationId As Long) As AnnotationReply() Implements IAnnotationReplyDataHandler.GetReplies
            SyncLock _syncRoot
                Try
                    Return Data.Where(Function(x) x.AnnotationId = annotationId).OrderBy(Function(r) r.RepliedOn).ToArray()
                Catch e As Exception
                    Throw New DataJsonException("Failed to get annotation replies.", e)
                End Try
            End SyncLock
        End Function

        Public Function DeleteReplyAndChildReplies(replyId As Long) As Boolean Implements IAnnotationReplyDataHandler.DeleteReplyAndChildReplies
            SyncLock _syncRoot
                Try
                    Dim data = Me.Data
                    Dim returnValue As Boolean = True

                    Dim reply As AnnotationReply = data.FirstOrDefault(Function(r) r.Id = replyId)
                    Dim childReplies As New List(Of AnnotationReply)()
                    Dim repliesOfCurrentLevel As AnnotationReply() = data.Where(Function(r) r.ParentReplyId = replyId).ToArray()
                    Dim repliesOfNextLevel As New List(Of AnnotationReply)()

                    While repliesOfCurrentLevel.Length > 0
                        repliesOfNextLevel.Clear()
                        childReplies.AddRange(repliesOfCurrentLevel)

                        For i As Integer = 0 To repliesOfCurrentLevel.Length - 1
                            Dim id As Decimal = repliesOfCurrentLevel(i).Id
                            repliesOfNextLevel.AddRange(data.Where(Function(r) r.ParentReplyId = id).ToArray())
                        Next

                        repliesOfCurrentLevel = repliesOfNextLevel.ToArray()
                    End While

                    childReplies.Reverse()
                    childReplies.ForEach(Function(x) returnValue = returnValue And MyBase.Remove(x))

                    returnValue = returnValue And MyBase.Remove(reply)
                    Return returnValue
                Catch e As Exception
                    Throw New DataJsonException("Failed to delete annotation replies.", e)
                End Try
            End SyncLock
        End Function
    End Class
End Namespace

File: Examples/GroupDocs.Annotation.Examples.CSharp/GroupDocs.AnnotationDataStorage/Repositories/AnnotationReplyRepository.cs
using System;
using System.Collections.Generic;
using System.Linq; 
using GroupDocs.Annotation.Handler.Input.DataObjects;
using GroupDocs.Annotation.Handler.Input;
namespace GroupDocs.Data.Json.Repositories
{
    public class AnnotationReplyRepository : JsonRepository<AnnotationReply>, IAnnotationReplyDataHandler
    {
        
        private const string _repoName = "GroupDocs.annotation.replies.json";

        public AnnotationReplyRepository(IRepositoryPathFinder pathFinder)
            : base(pathFinder.Find(_repoName))
        {
        }

        public AnnotationReplyRepository(string filePath)
            : base(filePath)
        {
        }

        public AnnotationReply GetReply(string guid)
        {
            lock (_syncRoot)
            {
                try
                {
                    return Data.FirstOrDefault(x => x.Guid == guid);
                }
                catch (Exception e)
                {
                    throw new DataJsonException("Failed to get annotation reply by GUID.", e);
                }
            }
        }

        public AnnotationReply[] GetReplies(long annotationId)
        {
            lock (_syncRoot)
            {
                try
                {
                    return Data
                        .Where(x => x.AnnotationId == annotationId)
                        .OrderBy(r => r.RepliedOn)
                        .ToArray();
                }
                catch (Exception e)
                {
                    throw new DataJsonException("Failed to get annotation replies.", e);
                }
            }
        }

        public bool DeleteReplyAndChildReplies(long replyId)
        {
            lock (_syncRoot)
            {
                try
                {
                    var data = this.Data;
                    bool returnValue = true;

                    AnnotationReply reply = data.FirstOrDefault(r => r.Id == replyId);
                    List<AnnotationReply> childReplies = new List<AnnotationReply>();
                    AnnotationReply[] repliesOfCurrentLevel = data.Where(r => r.ParentReplyId == replyId).ToArray();
                    List<AnnotationReply> repliesOfNextLevel = new List<AnnotationReply>();

                    while (repliesOfCurrentLevel.Length > 0)
                    {
                        repliesOfNextLevel.Clear();
                        childReplies.AddRange(repliesOfCurrentLevel);

                        for (int i = 0; i < repliesOfCurrentLevel.Length; i++)
                        {
                            decimal id = repliesOfCurrentLevel[i].Id;
                            repliesOfNextLevel.AddRange(
                                data.Where(r => r.ParentReplyId == id).ToArray());
                        }

                        repliesOfCurrentLevel = repliesOfNextLevel.ToArray();
                    }

                    childReplies.Reverse();
                    childReplies.ForEach(x => returnValue &= base.Remove(x));

                    returnValue &= base.Remove(reply);
                    return returnValue;
                }
                catch (Exception e)
                {
                    throw new DataJsonException("Failed to delete annotation replies.", e);
                }
            }
        }
    }
}
